In [1]:
import pandas as pd 
pd.set_option('display.max_columns', None)
import numpy as np
import os

import plotly.express as px
import plotly.graph_objects as go




In [2]:
filenames = ["data/msn_02_fuel_leak_signals_preprocessed.csv", "data/msn_10_fuel_leak_signals_preprocessed.csv", "data/msn_11_fuel_leak_signals_preprocessed.csv", "data/msn_12_fuel_leak_signals_preprocessed.csv", "data/msn_14_fuel_leak_signals_preprocessed.csv", "data/msn_29_fuel_leak_signals_preprocessed.csv", "data/msn_37_fuel_leak_signals_preprocessed.csv", "data/msn_53_fuel_leak_signals_preprocessed.csv"]
valid_cols = ['UTC_TIME', 'FUEL_USED_2', 'FUEL_USED_3', 'FUEL_USED_4',
       'FW_GEO_ALTITUDE', 'VALUE_FOB', 'VALUE_FUEL_QTY_CT',
       'VALUE_FUEL_QTY_FT1', 'VALUE_FUEL_QTY_FT2', 'VALUE_FUEL_QTY_FT3',
       'VALUE_FUEL_QTY_FT4', 'VALUE_FUEL_QTY_LXT', 'VALUE_FUEL_QTY_RXT',
       'FLIGHT_PHASE_COUNT', 'FUEL_USED_1', 'Flight', 'MSN']
datasets = []

for filename in filenames:
    datasets.append(pd.read_csv(filename, sep=";"))


In [2]:
data = pd.read_csv("data/msn_02_fuel_leak_signals_preprocessed.csv", sep=";")
data.shape

(623580, 111)

In [3]:
def Show_me_the_basics(df):
    resdf = pd.DataFrame()
    resdf['Column'] = df.columns
    resdf['Type'] = df.dtypes.values
    resdf['Rows'] = df.shape[0]
    resdf['# NaNs'] = df.isna().sum().values
    resdf['% NaNs'] = resdf['# NaNs'] / resdf['Rows'] *100
    resdf['# Uniques'] = df.nunique().values
    resdf['Uniques'] = [list(df[col].unique()) if df[col].nunique() < 13 else 'Too many to list' for col in df.columns]
    return resdf

In [4]:
data.UTC_TIME = pd.to_datetime(data.UTC_TIME)

In [5]:
nansdata = Show_me_the_basics(data)
#pd.set_option('display.max_rows', None)
nansdata

Column            Type    Rows  # NaNs  \
0                                UTC_TIME  datetime64[ns]  623580       0   
1                                     MSN          object  623580       0   
2                                  Flight          object  623580       0   
3                        ENGINE_RUNNING_1            bool  623580       0   
4                        ENGINE_RUNNING_2            bool  623580       0   
5                        ENGINE_RUNNING_3            bool  623580       0   
6                        ENGINE_RUNNING_4            bool  623580       0   
7                      FLIGHT_PHASE_COUNT         float64  623580   62922   
8                             FUEL_FLOW_1         float64  623580  108643   
9                             FUEL_FLOW_2         float64  623580  106986   
10                            FUEL_FLOW_3         float64  623580  106981   
11                            FUEL_FLOW_4         float64  623580  108869   
12                             FUEL_PITCH         float64  623580   63266   
13                              FUEL_ROLL         float64  623580   63266   
14               FUEL_TRANSFER_MODE_VALUE         float64  623580   63214   
15                            FUEL_USED_1         float64  623580  108636   
16                            FUEL_USED_2         float64  623580  106980   
17                            FUEL_USED_3         float64  623580  106979   
18                            FUEL_USED_4         float64  623580  108867   
19                        FW_GEO_ALTITUDE         float64  623580   62875   
20               LEAK_DETECTION_LEAK_FLOW         float64  623580   63275   
21                         LSTU1_A_VOLUME         float64  623580   63253   
22                         LSTU1_F_VOLUME         float64  623580   63253   
23                            PITCH_ANGLE         float64  623580   62850   
24                             ROLL_ANGLE         float64  623580   62851   
25                         RSTU1_A_VOLUME         float64  623580   63253   
26                         RSTU1_F_VOLUME         float64  623580   63253   
27          SELECTED_GADIR_ALTITUDE_VALUE         float64  623580   63289   
28            STATE_PMP_MAIN_FT1_ABNRM_ON            bool  623580       0   
29            STATE_PMP_MAIN_FT1_IMMERSED            bool  623580       0   
30                  STATE_PMP_MAIN_FT1_ON            bool  623580       0   
31            STATE_PMP_MAIN_FT2_ABNRM_ON            bool  623580       0   
32            STATE_PMP_MAIN_FT2_IMMERSED            bool  623580       0   
33                  STATE_PMP_MAIN_FT2_ON            bool  623580       0   
34            STATE_PMP_MAIN_FT3_ABNRM_ON            bool  623580       0   
35            STATE_PMP_MAIN_FT3_IMMERSED            bool  623580       0   
36                  STATE_PMP_MAIN_FT3_ON            bool  623580       0   
37            STATE_PMP_MAIN_FT4_ABNRM_ON            bool  623580       0   
38            STATE_PMP_MAIN_FT4_IMMERSED            bool  623580       0   
39                  STATE_PMP_MAIN_FT4_ON            bool  623580       0   
40            STATE_PMP_STBY_FT1_ABNRM_ON            bool  623580       0   
41            STATE_PMP_STBY_FT1_IMMERSED            bool  623580       0   
42                  STATE_PMP_STBY_FT1_ON            bool  623580       0   
43            STATE_PMP_STBY_FT2_ABNRM_ON            bool  623580       0   
44            STATE_PMP_STBY_FT2_IMMERSED            bool  623580       0   
45                  STATE_PMP_STBY_FT2_ON            bool  623580       0   
46            STATE_PMP_STBY_FT3_ABNRM_ON            bool  623580       0   
47            STATE_PMP_STBY_FT3_IMMERSED            bool  623580       0   
48                  STATE_PMP_STBY_FT3_ON            bool  623580       0   
49            STATE_PMP_STBY_FT4_ABNRM_ON            bool  623580       0   
50            STATE_PMP_STBY_FT4_IMMERSED            bool  623580       0   
51                  STATE_PMP_S

### Lets add some features

### ANALYZING FLIGHTS

In [7]:
def flight_basics(df):
    flightnumls = list(df.Flight.unique())
    resdfls = []
    for flightnum in flightnumls:
        df_flight = df[df.Flight == flightnum]
        df_flight_basics = Show_me_the_basics(df_flight)
        resdfls.append((flightnum, df_flight_basics))

    return resdfls



In [16]:
flight_basic_data = flight_basics(data)
flight_basic_data[0]

('V0136',
                                    Column            Type   Rows  # NaNs  \
 0                                UTC_TIME  datetime64[ns]  16934       0   
 1                                     MSN          object  16934       0   
 2                                  Flight          object  16934       0   
 3                        ENGINE_RUNNING_1            bool  16934       0   
 4                        ENGINE_RUNNING_2            bool  16934       0   
 5                        ENGINE_RUNNING_3            bool  16934       0   
 6                        ENGINE_RUNNING_4            bool  16934       0   
 7                      FLIGHT_PHASE_COUNT         float64  16934       1   
 8                             FUEL_FLOW_1         float64  16934       1   
 9                             FUEL_FLOW_2         float64  16934       1   
 10                            FUEL_FLOW_3         float64  16934       1   
 11                            FUEL_FLOW_4         float64  16934 

In [6]:
def make_flights_dfs_ls(df):
    flightnumls = list(df.Flight.unique())
    resdfls = []
    for flightnum in flightnumls:
        df_flight = df[df.Flight == flightnum]
        resdfls.append((flightnum,df_flight))
    return resdfls

In [8]:
dataperflight = make_flights_dfs_ls(data)

In [19]:

flightAltitudeescriptives = pd.DataFrame(columns=['Flight', 'count', 'mean', 'std', 'min', '25%', '50%', '75%', 'max', ])

for flight, planedata in dataperflight:
    describeSeries = planedata.FW_GEO_ALTITUDE.describe()
    flightAltitudeescriptives = flightAltitudeescriptives.append({'Flight': flight, 
                                      'count': describeSeries['count'], 
                                      'mean': describeSeries['mean'], 'std': describeSeries['std'], 
                                      'min': describeSeries['min'], '25%': describeSeries['25%'], 
                                      '50%': describeSeries['50%'], '75%': describeSeries['75%'], 
                                      'max': describeSeries['max'],
                                        
                                        }, 
                                          ignore_index=True)
    
flightAltitudeescriptives

    

Flight    count          mean           std         min           25%  \
0   V0136  16933.0  13694.095037  10695.967887   247.62500   3223.125000   
1   V0133  14322.0  12373.932404  11093.171610   248.87500   3011.687500   
2   V0926   9154.0   3290.233113   3322.321754   534.68020    541.000000   
3   V0837  18890.0   6606.764068   5067.151006   515.29420    542.106275   
4   V0626  15178.0    750.230055    506.548976   493.04790    503.713125   
5   V0929  19292.0  13603.932474   8734.108658    83.71345   7615.496500   
6   V0095  19429.0   8426.912709   5613.799609   259.00000   4275.125000   
7   V0623  14871.0  17552.542622  12447.037082    85.87821   3059.273000   
8   V0965  20826.0   8976.836524   9037.525004    86.05330   3708.689750   
9   V0344  11488.0  14140.174656   8920.751591   507.16740   1889.887750   
10  V0346  10113.0   2804.529073   3017.205383   511.94340    529.889100   
11  V0835   8662.0  18714.065220  12624.515924    94.91450   4022.094500   
12  V0624   7996.0  15495.175610  11942.731671    87.91195    516.079725   
13  V0097  22165.0  15240.236372  13599.547810     0.00000    273.250000   
14  V0962  16100.0   3751.054150   4297.161990  -709.89640    539.463200   
15  V0098  15064.0  12639.718878  10717.853595   250.50000   2090.781250   
16  V0134   9488.0  13447.871635   9159.697793   241.50000    265.125000   
17  V0889  26946.0  17688.316380  14026.140347    89.83078    104.419600   
18  V0348   8050.0  15695.405990  13673.169470   113.42880    528.677825   
19  V0140  13506.0  16770.798804  13100.944245   250.75000    275.687500   
20  V0135  20552.0  12232.790196   7512.963578   244.50000   8203.000000   
21  V0099  17374.0   7678.105910   4636.011255   252.50000   5583.343750   
22  V0925  13681.0  17945.955002  10963.118061   538.81990   8062.671000   
23  V0834   9881.0  25453.645394  14108.143829    88.17352  13956.550000   
24  V0625  15480.0   2243.046939   1371.637048   192.68090    846.536200   
25  V0963  19290.0  18455.149344  12113.210588 -1097.59200   5242.178750   
26  V0886   2249.0    530.461986    718.629276    88.33945    265.949700   
27  V0966  17860.0   8377.912870   8256.743535    82.37575   2315.964250   
28  V0347   5272.0   1054.790546   1168.231999   513.72010    520.873750   
29  V0833   8773.0   4479.249968   3914.375877    94.29116    449.683500   
30  V0622  11890.0  16513.498957  11864.509908    97.14664    925.506125   
31  V0964  13463.0  13198.701178  12664.090720    76.84367   5027.784000   
32  V0884   6092.0   1965.172540   2074.023307   280.30970   1068.385000   
33  V0138  12869.0    433.950647    320.761923   161.00000    263.500000   
34  V0345  14703.0   1935.739217   1550.518667   500.80440    514.616900   
35  V0890   7054.0    154.799198    215.863005    90.58122    101.000325   
36  V0836   9056.0  13622.336254  13246.871568    87.80524    168.555775   
37  V0928   9794.0   7682.151288   6481.058526    91.38740   1556.916250   
38  V0094  13154.0   9395.507819   6048.268076   257.12500   6531.593750   
39  V0132   6111.0   8189.008031   9184.825154   250.87500    262.125000   
40  V0927   7728.0  17713.786691  13982.782855    93.92255    547.944400   
41  V0137  19906.0  12398.017367   7161.135012   244.87500   8318.625000   

            50%           75%        max  
0   15238.88000  20429.250000  34395.000  
1   14177.94000  21048.095000  33084.250  
2    2228.30550   5044.696750  14324.730  
3    6295.35500   9960.344750  15349.200  
4     507.17895    517.761625   2209.664  
5   19455.71000  20604.120000  25810.080  
6    9511.25000  10794.630000  26437.000  
7   25069.66000  26874.445000  31212.190  
8    4848.64650  10296.892500  29080.000  
9   20632.95000  20659.650000  20712.400  
10   1283.96400   3078.900000  11154.400  
11  21263.67000  30329.230000  35485.620  
12  20566.07500  20657.395000  34905.700  
13  15684.38000  27423.630000  35481.380  
14   1576.96650   5934.848500  14629.500  
15  13232.69000  16186.410000  34688.880  
16 

In [22]:

for flight, planedata in dataperflight:
    planedata.UTC_TIME = pd.to_datetime(planedata.UTC_TIME)
    flightduration = pd.Timedelta(planedata.UTC_TIME.max() - planedata.UTC_TIME.min())
    flightAltitudeescriptives.loc[flightAltitudeescriptives.Flight == flight, 'FlightDuration'] = flightduration

flightAltitudeescriptives['FlightSeconds'] = flightAltitudeescriptives['FlightDuration'].dt.total_seconds()
flightAltitudeescriptives['RowsPerSecond'] = flightAltitudeescriptives['count'] / flightAltitudeescriptives['FlightSeconds']

c:\Users\niels\anaconda3\envs\data2\lib\site-packages\pandas\core\generic.py:5170: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [70]:
def make_IQR_outlier_bound(descriptivesdf):
    descriptivesdf['IQR'] = descriptivesdf['75%'] - descriptivesdf['25%']
    descriptivesdf['UpperBound'] = descriptivesdf['75%'] + 1.5 * descriptivesdf['IQR']
    descriptivesdf['LowerBound'] = descriptivesdf['25%'] - 1.5 * descriptivesdf['IQR']
    return descriptivesdf

In [71]:
flightAltitudeescriptives = make_IQR_outlier_bound(flightAltitudeescriptives)
flightAltitudeescriptives.sort_values(by='FlightDuration', ascending=False)

Flight    count          mean           std         min           25%  \
35  V0890   7054.0    154.799198    215.863005    90.58122    101.000325   
41  V0137  19906.0  12398.017367   7161.135012   244.87500   8318.625000   
20  V0135  20552.0  12232.790196   7512.963578   244.50000   8203.000000   
5   V0929  19292.0  13603.932474   8734.108658    83.71345   7615.496500   
13  V0097  22165.0  15240.236372  13599.547810     0.00000    273.250000   
8   V0965  20826.0   8976.836524   9037.525004    86.05330   3708.689750   
6   V0095  19429.0   8426.912709   5613.799609   259.00000   4275.125000   
25  V0963  19290.0  18455.149344  12113.210588 -1097.59200   5242.178750   
3   V0837  18890.0   6606.764068   5067.151006   515.29420    542.106275   
27  V0966  17860.0   8377.912870   8256.743535    82.37575   2315.964250   
21  V0099  17374.0   7678.105910   4636.011255   252.50000   5583.343750   
0   V0136  16933.0  13694.095037  10695.967887   247.62500   3223.125000   
14  V0962  16100.0   3751.054150   4297.161990  -709.89640    539.463200   
24  V0625  15480.0   2243.046939   1371.637048   192.68090    846.536200   
4   V0626  15178.0    750.230055    506.548976   493.04790    503.713125   
15  V0098  15064.0  12639.718878  10717.853595   250.50000   2090.781250   
7   V0623  14871.0  17552.542622  12447.037082    85.87821   3059.273000   
34  V0345  14703.0   1935.739217   1550.518667   500.80440    514.616900   
1   V0133  14322.0  12373.932404  11093.171610   248.87500   3011.687500   
22  V0925  13681.0  17945.955002  10963.118061   538.81990   8062.671000   
19  V0140  13506.0  16770.798804  13100.944245   250.75000    275.687500   
17  V0889  26946.0  17688.316380  14026.140347    89.83078    104.419600   
31  V0964  13463.0  13198.701178  12664.090720    76.84367   5027.784000   
38  V0094  13154.0   9395.507819   6048.268076   257.12500   6531.593750   
33  V0138  12869.0    433.950647    320.761923   161.00000    263.500000   
30  V0622  11890.0  16513.498957  11864.509908    97.14664    925.506125   
9   V0344  11488.0  14140.174656   8920.751591   507.16740   1889.887750   
10  V0346  10113.0   2804.529073   3017.205383   511.94340    529.889100   
23  V0834   9881.0  25453.645394  14108.143829    88.17352  13956.550000   
37  V0928   9794.0   7682.151288   6481.058526    91.38740   1556.916250   
16  V0134   9488.0  13447.871635   9159.697793   241.50000    265.125000   
2   V0926   9154.0   3290.233113   3322.321754   534.68020    541.000000   
36  V0836   9056.0  13622.336254  13246.871568    87.80524    168.555775   
29  V0833   8773.0   4479.249968   3914.375877    94.29116    449.683500   
11  V0835   8662.0  18714.065220  12624.515924    94.91450   4022.094500   
18  V0348   8050.0  15695.405990  13673.169470   113.42880    528.677825   
12  V0624   7996.0  15495.175610  11942.731671    87.91195    516.079725   
40  V0927   7728.0  17713.786691  13982.782855    93.92255    547.944400   
39  V0132   6111.0   8189.008031   9184.825154   250.87500    262.125000   
32  V0884   6092.0   1965.172540   2074.023307   280.30970   1068.385000   
28  V0347   5272.0   1054.790546   1168.231999   513.72010    520.873750   
26  V0886   2249.0    530.461986    718.629276    88.33945    265.949700   

            50%           75%        max  FlightDuration  FlightSeconds  \
35    104.06145    105.944325   1250.526 0 days 13:05:36        47136.0   
41  14800.56500  17860.097500  24893.880 0 days 07:46:15        27975.0   
20  14090.56500  20272.562500  21313.250 0 days 07:38:58        27538.0   
5   19455.71000  20604.120000  25810.080 0 days 07:27:31        26851.0   
13  15684.38000  27423.630000  35481.380 0 days 06:09:25        22165.0   
8    4848.64650  10296.892500  29080.000 0 days 05:47:06        20826.0   
6    9511.25000  10794.630000  26437.000 0 days 05:23:49        19429.0   
25  22763.95000  25922.325000  34996.070 0 days 05:21:30        19290.0   
3    6295.35500   9960.344750  15349.200 0 days 05:14:50   

#### Pick a specific flight

In [26]:
def pick_specific_flight(df, flightnum):
    singleflightdata = df[df.Flight == flightnum]
    print(f'\n YOU ARE LOOKING AT FLIGHT: {flightnum} \n')
    return singleflightdata

Flights I Checked-out trying to understand flights and when they are anomalous. 

- V0889: The flight has 2 times the rows than it has seconds, so i had a closer look. We just have duplicate timestamps We can just drop duplicates --> UTC TIME has duplicates. 


- V0890: This flight only has rows for 14% of its duration. here we need to dropna() theres a whole bunch of hours where we have time readings but no data. The data could then be visualized, however we see that altitude changes irregularly. we need to chek a little more to understand whats going on. We see the value_fob deceasing and an empty central tank. the LX and RX tank show irregular behavior that combined with the behvaior of the altitude makes me think that the plane has either in the landing (which is still strange given the 1.30h period in which altitude was 0) or the plane's fuel tanks are being emptied. For this i will have a closer look at engine data. Research describes the steps to the process of emptying fuel tanks as: Fuel Transfer, Fuel Drainage, Fuel Consumption, Fuel Venting. These steps are conducted generally during maintenance and fuel tank inspection. For this the engines are therefore turned on but the plane is on the ground. The way LXT and RXT dump fuel in the graph indicate to me the system is not pressurized. Looking at the engineered features said it all. we can see the rate at which value FOB is dissapearing is per second geater than the rate at which the engines are burning fuel, and this delta is increasing in size. 

So what characteized this maintenace check: Altitude behvaing irregularly, mainly staying below 1250 (which is its outlier, Upper IR bound 113). The DELTA_VFOBM_VS_SFUSED is increasing in size, meaning the rate at which fuel is going missing from Value_Fobs first instance is steadily greate than the sum of fuel used by the engines.

- V0929: This flight has rows only for 71% of its duration. Again we see data only after a long period of nan values, we therefore dropped nan values We see drastic changes in altitude in the duration of the flight, the plane dives to low altitude then increases its altitude again


- V0137: FIGHT GOING UP AND DOWN, NORMAL LOOKING TANK BEHVAIOR (TEST FLIGHT)

- V0135:

- V0624:

- V0835:

- V0097
- V0095
- V0138
- V0622
- V0925
- V0344 - NORMAL LOOKING ALTITUDE
- V0346 - SPIKE TOUCH AND GOES
- V0140 - Flight goes up doown then up again.
- V0925 - SPIKE Normal looking altitude
- V0926 - SPIKE BIG UP AND DOWN IN ALTITUDE 
- V0927 - 

MOST HAVE THIS STRANGE SPIKE IN FUEL USED WHEN TURNING ON THE ENGINE (WE SHOULD CHECK THE FLIGHT PHASE OF THIS). looks like its an effect of phase 1 and just spinning up the engines. Not all have this issue

In [459]:
flightIwant = "V0927"
singleflightdata = pick_specific_flight(data, flightIwant)


 YOU ARE LOOKING AT FLIGHT: V0926 



In [460]:
colsOfIntrest = ['UTC_TIME','FUEL_USED_1', 'FUEL_USED_2', 'FUEL_USED_3', 'FUEL_USED_4',
         'FW_GEO_ALTITUDE', 'VALUE_FOB', 'VALUE_FUEL_QTY_CT', 'VALUE_FUEL_QTY_FT1', 
            'VALUE_FUEL_QTY_FT2', 'VALUE_FUEL_QTY_FT3', 'VALUE_FUEL_QTY_FT4', 'VALUE_FUEL_QTY_LXT',
            'VALUE_FUEL_QTY_RXT', 'FLIGHT_PHASE_COUNT', 'Flight', 'MSN']

singleflightdata = singleflightdata[colsOfIntrest]

In [461]:
singleflightdata.shape

(9155, 17)

In [462]:
singleflightdata.dropna(inplace=True)

In [463]:
singleflightdata.FLIGHT_PHASE_COUNT = singleflightdata.FLIGHT_PHASE_COUNT.floordiv(1)

In [464]:
singleflightdata.shape

(6734, 17)

In [465]:
# SMOOTH

singleflightdata['UTC_TIME'] = pd.to_datetime(singleflightdata['UTC_TIME'])
singleflightdata.set_index('UTC_TIME', inplace=True)
singleflightdata = singleflightdata.resample('5S').mean()
singleglightdata = singleflightdata.interpolate(method='linear', limit_direction='both')
singleflightdata.reset_index(inplace=True)


In [466]:
singleflightdata.shape

(1348, 15)

In [467]:
# FEATURE ENGINEERING

singleflightdata['SUM_FUEL_USED'] = singleflightdata['FUEL_USED_1'] + singleflightdata['FUEL_USED_2'] + singleflightdata['FUEL_USED_3'] + singleflightdata['FUEL_USED_4']
singleflightdata['SUM_FUEL_QTY'] = singleflightdata['VALUE_FUEL_QTY_CT'] + singleflightdata['VALUE_FUEL_QTY_FT1'] + singleflightdata['VALUE_FUEL_QTY_FT2'] + singleflightdata['VALUE_FUEL_QTY_FT3'] + singleflightdata['VALUE_FUEL_QTY_FT4'] + singleflightdata['VALUE_FUEL_QTY_LXT'] + singleflightdata['VALUE_FUEL_QTY_RXT']
singleflightdata['VALUE_FOB_MISSING'] = singleflightdata['VALUE_FOB'].iloc[0] - singleflightdata['VALUE_FOB']
singleflightdata['DELTA_VFOB_VS_SQTY'] = singleflightdata['VALUE_FOB'] - singleflightdata['SUM_FUEL_QTY']
singleflightdata['DELTA_VFOBM_VS_SFUSED'] = singleflightdata['VALUE_FOB_MISSING'] - singleflightdata['SUM_FUEL_USED']

singleflightdata['ALTITUDE_DIFF'] = singleflightdata['FW_GEO_ALTITUDE'].diff()


In [468]:
phase_dict = {
    1: 'Pre-flight',
    2: 'Engine Run',
    3: 'Take-Off 1',
    4: 'Take-Off 2',
    5: 'Take-Off 3',
    6: 'Climbing 1',
    7: 'Climbing 2',
    8: 'Cruise',
    9: 'Descent',
    10: 'Approach',
    11: 'Landing',
    12: 'Post-flight'
}

phaselabels = [phase_dict.get(phase, 'Unknown') for phase in singleflightdata['FLIGHT_PHASE_COUNT']]

In [469]:
fig = go.Figure()

fig.add_trace(go.Pie(labels=phaselabels, values=singleflightdata['FLIGHT_PHASE_COUNT']))
fig.update_layout(title='FLIGHT_PHASE_COUNT', width=800, height=500, template='plotly_dark')

In [470]:
fig = go.Figure()

fig.add_trace(go.Scatter(x=singleflightdata.UTC_TIME, y=singleflightdata.SUM_FUEL_USED, mode='lines', name='Fuel Used' ))
fig.add_trace(go.Scatter(x=singleflightdata.UTC_TIME, y=singleflightdata.SUM_FUEL_QTY, mode='lines', name='Fuel Quantity' ))
fig.add_trace(go.Scatter(x=singleflightdata.UTC_TIME, y=singleflightdata.VALUE_FOB_MISSING, mode='lines', name='Value FOB Missing' ))
fig.add_trace(go.Scatter(x=singleflightdata.UTC_TIME, y=singleflightdata.VALUE_FOB, mode='lines', name='Value FOB' ))

fig.update_layout(title=f'Flight {flightIwant} Engineered Features', template='plotly_dark')


In [471]:
fig = go.Figure()

fig.add_trace(go.Scatter(x=singleflightdata.UTC_TIME, y=singleflightdata.ALTITUDE_DIFF, mode='lines', name='ALTITUDE DIFF' ))

fig.update_layout(title=f'Flight {flightIwant} Engineered Features', template='plotly_dark')


In [472]:
#phase_labels = [f'Phase {int(i)}' for i in singleflightdata.FLIGHT_PHASE_COUNT]

#fig = go.Figure()

#fig.add_trace(go.Scatter(
#   x=singleflightdata.UTC_TIME,
#    y=singleflightdata.ALTITUDE_DIFF,
#    mode='lines',
#    name='ALTITUDE DIFF',
#    line=dict(color=phase_labels, colorscale='Viridis', showscale=True)
#))

#fig.update_layout(
#    title=f'Flight {flightIwant} Engineered Features',
#    template='plotly_dark'
#)


In [473]:
fig = go.Figure()

fig.add_trace(go.Scatter(x=singleflightdata.UTC_TIME, y=singleflightdata.DELTA_VFOB_VS_SQTY, mode='lines', name='Delta VFOB vs SQTY' ))
fig.add_trace(go.Scatter(x=singleflightdata.UTC_TIME, y=singleflightdata.DELTA_VFOBM_VS_SFUSED, mode='lines', name='Delta VFOBM vs SFUSED' ))

fig.update_layout(title=f'Flight {flightIwant} Engineered Features', template='plotly_dark')


In [474]:
fig = go.Figure()

fig.add_trace(go.Scatter(x=singleflightdata.UTC_TIME, y=singleflightdata.FW_GEO_ALTITUDE, mode='lines', name='Altitude'))
fig.add_trace(go.Scatter(x=singleflightdata.UTC_TIME, y=singleflightdata.VALUE_FOB, mode='lines', name='Value FOB' ))

fig.update_layout(title=f'Flight {flightIwant} Altitude and Value FOB', template='plotly_dark')

In [475]:
fig = go.Figure()

fig.add_trace(go.Scatter(x=singleflightdata.UTC_TIME, y=singleflightdata.VALUE_FUEL_QTY_CT, mode='lines', name='Central Tank'))
fig.add_trace(go.Scatter(x=singleflightdata.UTC_TIME, y=singleflightdata.VALUE_FUEL_QTY_LXT, mode='lines', name='Left Tank' ))
fig.add_trace(go.Scatter(x=singleflightdata.UTC_TIME, y=singleflightdata.VALUE_FUEL_QTY_RXT, mode='lines', name='Right Tank' ))

fig.update_layout(title=f'Flight {flightIwant} Fuel Quantities', template='plotly_dark')

In [476]:
fig = go.Figure()

fig.add_trace(go.Scatter(x=singleflightdata.UTC_TIME, y=singleflightdata.VALUE_FUEL_QTY_FT1, mode='lines', name='Fuel Tank 1'))
fig.add_trace(go.Scatter(x=singleflightdata.UTC_TIME, y=singleflightdata.VALUE_FUEL_QTY_FT2, mode='lines', name='Fuel Tank 2' ))
fig.add_trace(go.Scatter(x=singleflightdata.UTC_TIME, y=singleflightdata.VALUE_FUEL_QTY_FT3, mode='lines', name='Fuel Tank 3' ))
fig.add_trace(go.Scatter(x=singleflightdata.UTC_TIME, y=singleflightdata.VALUE_FUEL_QTY_FT4, mode='lines', name='Fuel Tank 4' ))

fig.update_layout(title=f'Flight {flightIwant} Fuel Quantities', template='plotly_dark')

In [477]:
fig = go.Figure()

fig.add_trace(go.Scatter(x=singleflightdata.UTC_TIME, y=singleflightdata.FUEL_USED_1, mode='lines', name='Fuel Used 1'))
fig.add_trace(go.Scatter(x=singleflightdata.UTC_TIME, y=singleflightdata.FUEL_USED_2, mode='lines', name='Fuel Used 2' ))
fig.add_trace(go.Scatter(x=singleflightdata.UTC_TIME, y=singleflightdata.FUEL_USED_3, mode='lines', name='Fuel Used 3' ))
fig.add_trace(go.Scatter(x=singleflightdata.UTC_TIME, y=singleflightdata.FUEL_USED_4, mode='lines', name='Fuel Used 4' ))

fig.update_layout(title=f'Flight {flightIwant} Fuel Used', template='plotly_dark')

In [478]:
singleflightdata.head(25)[['FLIGHT_PHASE_COUNT', 'UTC_TIME', 'ALTITUDE_DIFF', 'VALUE_FOB', 'FUEL_USED_1', 'FUEL_USED_2', 'FUEL_USED_3' , 'FUEL_USED_4']]

FLIGHT_PHASE_COUNT            UTC_TIME  ALTITUDE_DIFF     VALUE_FOB  \
0                  1.0 2016-10-07 07:36:15            NaN  17930.333333   
1                  1.0 2016-10-07 07:36:20      -0.127107  17929.000000   
2                  1.0 2016-10-07 07:36:25      -0.425440  17929.400000   
3                  1.0 2016-10-07 07:36:30      -0.526660  17932.600000   
4                  1.0 2016-10-07 07:36:35       0.197640  17931.200000   
5                  1.0 2016-10-07 07:36:40       0.052540  17931.000000   
6                  1.0 2016-10-07 07:36:45      -0.248560  17929.200000   
7                  1.0 2016-10-07 07:36:50      -0.092760  17927.000000   
8                  1.0 2016-10-07 07:36:55       0.052680  17926.200000   
9                  1.0 2016-10-07 07:37:00      -0.009500  17927.800000   
10                 1.0 2016-10-07 07:37:05      -0.410660  17927.400000   
11                 1.0 2016-10-07 07:37:10      -0.747760  17926.000000   
12                 1.0 2016-10-07 07:37:15      -0.416740  17924.800000   
13                 1.0 2016-10-07 07:37:20      -0.151820  17922.600000   
14                 1.0 2016-10-07 07:37:25      -0.351520  17921.000000   
15                 1.0 2016-10-07 07:37:30      -0.144960  17918.200000   
16                 1.6 2016-10-07 07:37:35       0.166840  17917.200000   
17                 2.0 2016-10-07 07:37:40      -0.374440  17916.600000   
18                 2.0 2016-10-07 07:37:45       0.614260  17916.400000   
19                 2.0 2016-10-07 07:37:50       0.356860  17916.800000   
20                 2.0 2016-10-07 07:37:55       0.130280  17915.000000   
21                 2.0 2016-10-07 07:38:00       0.263320  17913.800000   
22                 2.0 2016-10-07 07:38:05       0.299940  17915.800000   
23                 2.0 2016-10-07 07:38:10       0.514500  17916.200000   
24                 2.0 2016-10-07 07:38:15       0.090920  17914.800000   

    FUEL_USED_1  FUEL_USED_2  FUEL_USED_3  FUEL_USED_4  
0   3827.901000  3811.325000  3795.018000     3845.375  
1   3827.901000  3811.325000  3795.018000     3845.375  
2   3827.901000  3811.325000  3795.018000     3845.375  
3      0.000000  3811.325000  3795.018000     3845.375  
4      0.000000  3811.325000  3795.018000     3845.375  
5      0.000525  3811.325000  3795.018000     3845.375  
6      0.001283  3811.325000  3795.018000     3845.375  
7      0.001372  3811.325000  3795.018000     3845.375  
8      0.001817  3811.325000  3795.018000     3845.375  
9      0.002452  3811.325000  3795.018000     3845.375  
10     0.002953  2286.795000  3795.018000     3845.375  
11     0.019057     0.000000  3795.018000     3845.375  
12     0.397229     0.000000  3795.018000     3845.375  
13     0.928977     0.001281  3795.018000     3845.375  
14     1.450470     0.001380  3795.018000     3845.375  
15     1.974942     0.001618  3795.018000     3845.375  
16     2.493568     0.002489  3795.018000     3845.375  
17     3.012774     0.002878  1518.007200     3845.375  
18     3.531840     0.008334     0.000000     3845.375  
19     4.047282     0.344986     0.000000     3845.375  
20     4.565058     0.875129     0.001454     3845.375  
21     5.076596     1.375743     0.001471     3845.375  
22     5.593273     1.883225     0.002168     3845.375  
23     6.103635     2.381165     0.002592     3076.300  
24     6.620632     2.883207     0.003292        0.000

#### Pick a random flight

In [9]:
import random
def pick_rand_flight(df):
    flightnum = random.choice(df.Flight.unique().tolist())
    flightdata = df[df.Flight == flightnum]
    print(f'\n YOU ARE LOOKING AT FLIGHT: {flightnum} \n')
    return flightdata


In [19]:
# BEWARE OF ThiS
singleflightdf = pick_rand_flight(data)


 YOU ARE LOOKING AT FLIGHT: V0926 



In [24]:
singleflightdf.FW_GEO_ALTITUDE.describe()

count     9154.000000
mean      3290.233113
std       3322.321754
min        534.680200
25%        541.000000
50%       2228.305500
75%       5044.696750
max      14324.730000
Name: FW_GEO_ALTITUDE, dtype: float64

In [13]:
flightdf.head()

UTC_TIME         MSN Flight  ENGINE_RUNNING_1  \
194861 2010-10-28 07:14:46  A400M-0002  V0097              True   
194862 2010-10-28 07:14:47  A400M-0002  V0097             False   
194863 2010-10-28 07:14:48  A400M-0002  V0097             False   
194864 2010-10-28 07:14:49  A400M-0002  V0097             False   
194865 2010-10-28 07:14:50  A400M-0002  V0097             False   

        ENGINE_RUNNING_2  ENGINE_RUNNING_3  ENGINE_RUNNING_4  \
194861              True              True              True   
194862             False             False             False   
194863             False             False             False   
194864             False             False             False   
194865             False             False             False   

        FLIGHT_PHASE_COUNT  FUEL_FLOW_1  FUEL_FLOW_2  FUEL_FLOW_3  \
194861                 NaN          NaN          NaN          NaN   
194862                 1.0          0.0          0.0          0.0   
194863                 1.0          0.0          0.0          0.0   
194864                 1.0          0.0          0.0          0.0   
194865                 1.0          0.0          0.0          0.0   

        FUEL_FLOW_4  FUEL_PITCH  FUEL_ROLL  FUEL_TRANSFER_MODE_VALUE  \
194861          NaN         NaN        NaN                       NaN   
194862          0.0    1.599987   0.580852                       0.0   
194863          0.0    1.599987   0.580852                       0.0   
194864          0.0    1.599987   0.580852                       0.0   
194865          0.0    1.599987   0.580852                       0.0   

        FUEL_USED_1  FUEL_USED_2  FUEL_USED_3  FUEL_USED_4  FW_GEO_ALTITUDE  \
194861          NaN          NaN          NaN          NaN              NaN   
194862          0.0          0.0          0.0          0.0          265.875   
194863          0.0          0.0          0.0          0.0          265.875   
194864          0.0          0.0          0.0          0.0          265.750   
194865          0.0          0.0          0.0          0.0          265.875   

        LEAK_DETECTION_LEAK_FLOW  LSTU1_A_VOLUME  LSTU1_F_VOLUME  PITCH_ANGLE  \
194861                       NaN             NaN             NaN          NaN   
194862                       0.0        40.80074        28.61984     0.571289   
194863                       0.0        40.70916        28.56454     0.571289   
194864                       0.0        40.70884        28.31597     0.571289   
194865                       0.0        40.69891        28.05474     0.571289   

        ROLL_ANGLE  RSTU1_A_VOLUME  RSTU1_F_VOLUME  \
194861         NaN             NaN             NaN   
194862    0.527344        28.50451        39.44913   
194863    0.527344        28.49164        39.33009   
194864    0.527344        28.50263        39.37023   
194865    0.527344        28.48396        39.40307   

        SELECTED_GADIR_ALTITUDE_VALUE  STATE_PMP_MAIN_FT1_ABNRM_ON  \
194861                            NaN                         True   
194862                         -181.0                        False   
194863                         -181.0                        False   
194864                         -181.0                        False   
194865                         -181.0                        False   

        STATE_PMP_MAIN_FT1_IMMERSED  STATE_PMP_MAIN_FT1_ON  \
194861                         True                   True   
194862                         True                  False   
194863                         True                  False   
194864                         True                  False   
194865                         True                  False   

        STATE_PMP_MAIN_FT2_ABNRM_ON  STATE_PMP_MAIN_FT2_IMMERSED  \
194861                         True                         True   
194862                        False                         True   
194863                        False                         True   
194864                       

In [14]:
// Make a dataframe showing the min, mean and max values for the geo alttude column for each flight_basic_data










SyntaxError: invalid syntax (<ipython-input-14-7cd13f2b7084>, line 1)

#### Standardized data